functions.php

```diff
@@ -0,0 +1,42 @@
+<?php
+function dbpull():array{
+    $connectionString = 'mysql:host=db; dbname=endangeredlist';
+    $dbUsername = 'root';
+    $dbPassword = 'password';
+    $db = new PDO($connectionString, $dbUsername, $dbPassword);
+    $db->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
+
+    $queryString = 'SELECT * FROM `animals`;';
+    $query = $db->prepare($queryString);
+    $query->execute();
+    $allresults = $query->fetchAll();
+    return $allresults;
+}
+$animals = dbpull();
+
+function displayAnimals(array $animals):string {
+    $animalsHTML = '';
+     foreach ($animals as $animal){
+
+         if(!is_array($animal) ||
+         !array_key_exists('species', $animal) ||
+         !array_key_exists('population', $animal) ||
+         !array_key_exists('country', $animal) ||
+         !array_key_exists('charity', $animal))
+         {
+             return '';
+         }
+
+         $animalsHTML .=
+             '<div class = animal>' .
+                '<li class = species>Species: ' . $animal['species'] . '</li>'.
+                '<li class = facts>Individuals remaining :' . $animal['population'] . '</li>' .
+                '<li class = facts>Habitat: ' . $animal['country'] . '</li>' .
+                '<li class = facts>Charity: <a href = "' . $animal['charity'] . '">Link</a></li>' .
+            '</div>';
+     }
+     return $animalsHTML;
```

Comment:
indentation is weird in this function - i recommend doing Code > Reformat code in phpstorm to fix the indents in the whole file